Fix Windows Command::env("PATH")

ChrisDenton opened the pull request:
Fixes #87859

rust-highfive commented:
r? @dtolnay

(rust-highfive has picked a reviewer for you, use r? to override)

Urgau commented:
I think this need's to be beta nominated because the [PR](https://github.com/rust-lang/rust/pull/85270) who introduce the bug is the the current beta.

dtolnay reviewed (changes requested):
The `impl Borrow<OsStr> for EnvKey` lower in this file needs to be removed. That impl was the root cause. The impl is wrong because the Ord and Eq impls have different behavior between OsStr and EnvKey, which violates:

> In particular `Eq`, `Ord` and `Hash` must be equivalent for borrowed and owned values: `x.borrow() == y.borrow()` should give the same result as `x == y`.

ChrisDenton commented:
Ok I've removed the borrow and changed `sys_common::process::CommandEnv` to work with `EnvKey` instead of `OsStr`/`OsString`. I wonder if it might ultimately be worth Windows using its own implementation of `CommandEnv`.

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 419902e413ddd62673f8e54c71e5164d862aee08 has been approved by `dtolnay`

<!-- @bors r=dtolnay 419902e413ddd62673f8e54c71e5164d862aee08 -->
<!-- homu: {"type":"Approved","sha":"419902e413ddd62673f8e54c71e5164d862aee08","approver":"dtolnay"} -->

bors commented:
:hourglass: Testing commit 419902e413ddd62673f8e54c71e5164d862aee08 with merge 03e3dd097a4919dea6607213cffdbf33e65c0121...
<!-- homu: {"type":"BuildStarted","head_sha":"419902e413ddd62673f8e54c71e5164d862aee08","merge_sha":"03e3dd097a4919dea6607213cffdbf33e65c0121"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3278287944)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/3278287944","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3278287944) [(plain)](https://github.com/rust-lang-ci/rust/commit/03e3dd097a4919dea6607213cffdbf33e65c0121/checks/3278287944/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

ChrisDenton commented:
The failure looks like a spurious timeout or something?

dtolnay commented:
@bors retry

bors commented:
:hourglass: Testing commit 419902e413ddd62673f8e54c71e5164d862aee08 with merge 32994af1891b398c863fe81080d08df924e0a069...
<!-- homu: {"type":"BuildStarted","head_sha":"419902e413ddd62673f8e54c71e5164d862aee08","merge_sha":"32994af1891b398c863fe81080d08df924e0a069"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3282845928)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/3282845928","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3282792737) [(plain)](https://github.com/rust-lang-ci/rust/commit/32994af1891b398c863fe81080d08df924e0a069/checks/3282792737/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

ChrisDenton commented:
Hm, this seems to have failed downloading. Third time lucky?

ehuss commented:
@bors retry

Failed to download action 'https://api.github.com/repos/actions/checkout/tarball/5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f'. Error: The operation was canceled.

ChrisDenton commented:
Yeah, remove needs either an EnvKey or else a Borrow that has the same ordering as EnvKey. I could limit the allocation to right before `self.vars.remove`? That would fix the immediate problem and shouldn't have too much impact.

ChrisDenton commented on library/std/src/sys/windows/process.rs:
```diff
@@ -100,6 +105,20 @@ impl PartialEq for EnvKey {
         }
     }
 }
+impl PartialOrd<str> for EnvKey {
+    fn partial_cmp(&self, other: &str) -> Option<cmp::Ordering> {
+        Some(self.cmp(&EnvKey::new(other)))
+    }
+}
+impl PartialEq<str> for EnvKey {
+    fn eq(&self, other: &str) -> bool {
+        if self.os_string.len() != other.len() {
+            false
+        } else {
+            self.cmp(&EnvKey::new(other)) == cmp::Ordering::Equal
+        }
+    }
+}
```
Good points. I could remove those implementations and maybe I should add a `#[cfg(not(windows))]` to `have_changed_path`? So as to guard against the possibility of it accidentally being used in the future, or at least until a longer term fix is made.
